### ISSUE TITLE
[Bug] PTODSL-vmi，vmula 的 accumulator 操作数错误绑定为 rhs

### ISSUE THREAD

**Issue by chenben09** [hw-native-sys/PTOAS#1374]
### Component

PTO Dialect / ODS (include/PTO/IR)

### Description

在 A5 平台上，使用如下 PTODSL VMI 接口：
result = pto.vmi.vmula(
    acc,
    lhs,
    rhs,
    predicate,
    pmode=pmode,
)
接口期望的语义是：
result = acc + lhs * rhs

但是当前运行结果显示，实际执行效果等价于：
result = rhs + lhs * rhs
因此，vmula 的第三个输入操作数 rhs 疑似在后端 lowering、指令选择、寄存器分配或 Simulator 指令解释阶段被错误地绑定为 destructive destination，而第一个输入操作数 acc 没有被用作初始 accumulator。

### Reproduction (minimal)

```markdown
@jit_kernel(name, params)
    def kernel(
        acc_ptr: pto.ptr(dtype, "gm"),
        lhs_ptr: pto.ptr(dtype, "gm"),
        rhs_ptr: pto.ptr(dtype, "gm"),
        out_ptr: pto.ptr(dtype, "gm"),
    ):
        acc_ub = pto.castptr(pto.i64(acc_ub_offset), pto.ptr(dtype, "ub"))
        lhs_ub = pto.castptr(pto.i64(lhs_ub_offset), pto.ptr(dtype, "ub"))
        rhs_ub = pto.castptr(pto.i64(rhs_ub_offset), pto.ptr(dtype, "ub"))
        out_ub = pto.castptr(pto.i64(out_ub_offset), pto.ptr(dtype, "ub"))
        pto.mte_gm_ub(acc_ptr, acc_ub, 0, nbytes, nburst=(1, 0, 0))
        pto.mte_gm_ub(lhs_ptr, lhs_ub, 0, nbytes, nburst=(1, 0, 0))
        pto.mte_gm_ub(rhs_ptr, rhs_ub, 0, nbytes, nburst=(1, 0, 0))
        # For all-off zeroing Cases, prefill the destination with a non-zero
        # source. A missing/partial write can therefore not be mistaken for
        # the expected all-zero result. The corresponding config Case uses
        # nonzero inputs to make this sentinel deterministic.
        if sentinel:
            pto.mte_gm_ub(acc_ptr, out_ub, 0, nbytes, nburst=(1, 0, 0))
        pto.set_flag("MTE2", "V", event_id=0)
        pto.wait_flag("MTE2", "V", event_id=0)
        acc = pto.vmi.vload(acc_ub, pto.const(0, dtype=pto.index), size=vl)
        lhs = pto.vmi.vload(lhs_ub, pto.const(0, dtype=pto.index), size=vl)
        rhs = pto.vmi.vload(rhs_ub, pto.const(0, dtype=pto.index), size=vl)
        if masked:
            if mask_kind == "group_prefix":
                predicate = pto.vmi.create_mask(active_lanes, size=vl, group=group)
            else:
                predicate = pto.vmi.create_mask(active_lanes, size=vl)
            result = pto.vmi.vmula(acc, lhs, rhs, predicate, pmode=pmode)
        else:
            predicate = pto.vmi.create_mask(active_lanes, size=vl)
            result = pto.vmi.vmula(acc, lhs, rhs, predicate, pmode=pmode)
        pto.vmi.vstore(result, out_ub, pto.const(0, dtype=pto.index))
        pto.set_flag("V", "MTE3", event_id=0)
        pto.wait_flag("V", "MTE3", event_id=0)
        pto.mte_ub_gm(out_ub, out_ptr, nbytes, nburst=(1, 0, 0))
```

### Expected behavior

编译通过+运行通过

### Actual behavior / error logs

```text
当前精度对比失败

对于VL=1，使用以下的数据

操作数	数值
acc	4.6123962
lhs	-3.7436867
rhs	-0.9364810

计算逻辑：result = acc + lhs * rhs

golden结果为：
acc + lhs * rhs
= 4.6123962 + (-3.7436867) * (-0.9364810)
≈ 8.1182877

实际结果为：
rhs + lhs * rhs
= -0.9364810 + (-3.7436867) * (-0.9364810)
≈ 2.5694105

失败的用例（可在测试框架复现）：
vmi_vmula_f32_vl1_g1_a1_n_nomask
vmi_vmula_f32_vl128_g1_a128_n_nomask
```

### Git commit

10883a8e7e84c9ee6198cb86c98e763d21faa334

### Host platform

Linux (aarch64)

### Target Ascend arch (if relevant)

a5

### PTOAS build level (if relevant)

None

**Comment by TelGome** [comment]
## 1. 初始干扰：OPPROF 捕获到了 `ZerosLike`

最初的 OPPROF 中：

```text
OPPROF_20260902112516_WPGMQJADCTIKUDTH/dump/object_dump.txt
```

记录的是：

```text
ZerosLike_4_BYTES_high_performance_1
```

原因是测试框架通过以下代码创建输出 tensor：

```python
torch.zeros(shape, dtype=torch_dtype, device="npu:0")
```

这会在 DUT 之前启动 CANN 内置的 `ZerosLike` kernel，`msprof op simulator` 因而可能只捕获该前置 kernel。该 OPPROF 中的指令不能用于判断 `vmi_vmula` 的行为。

修复方式是在 simulator 模式下使用 `torch.empty`，真机模式仍保留 `torch.zeros`，避免改变 masked/preserve 用例中未写 lane 的初始化行为：

```python
if os.environ.get("PTODSL_SIMULATOR_MODE") == "1":
    out = torch.empty(shape, dtype=torch_dtype, device="npu:0")
else:
    out = torch.zeros(shape, dtype=torch_dtype, device="npu:0")
```

`sim_dsl.sh` 显式设置：

```bash
export PTODSL_SIMULATOR_MODE=1
```

修改后，OPPROF 捕获到的 kernel 变为：

```text
vmi_vmula_vmi_vmula_f32_vl1_g1_a1_n_nomask_mix_aiv
```
## 2. 最终根因：device tensor 生命周期错误

修复前 `_alloc_host_inputs` 的逻辑为：

```python
arr = data.generate(spec)
tensor = arr.to(device="npu:0")
host_inputs.append((arr, tensor.data_ptr()))
```

这里只保存了：

- CPU `arr`；
- NPU tensor 的裸 `data_ptr()`。

没有保存 NPU `tensor` 对象本身。循环进入下一次迭代后，前一个 device tensor 失去 Python 引用，其内存可以被释放并被后续输入分配复用。最终传给 kernel 的 `acc_ptr` 成为悬空指针，并读到了后续 `rhs` 的内容。

这解释了修复前的现象：

```text
V0 = rhs
V1 = lhs
V2 = rhs
```

问题不在 `RV_VMULA` 的运算，也不在 Bisheng 的寄存器选择，而在测试框架传入了生命周期已经结束的 device tensor 裸指针。

## 3. 修复

保存 host tensor、device tensor 和裸指针三者：

```python
arr = data.generate(spec)
tensor = arr.to(device="npu:0")
host_inputs.append((arr, tensor, tensor.data_ptr()))
```

构造 launch 参数时：

```python
launch_args = [ptr for _, _, ptr in host_inputs] + list(out_ptrs)
```

只要 `host_inputs` 仍在作用域中，device tensor 就保持存活，裸指针在 kernel launch 和 `torch.npu.synchronize()` 完成前始终有效。

失败证据落盘也同步适配三元组，修复了 `inputs.npz` 生成时错误解包并被 best-effort 异常处理静默吞掉的问题。

涉及文件：

```text
tools/ptodsl-test/lib/runtime.py
tools/ptodsl-test/scripts/sim_dsl.sh
```

## 4. 修复后验证

修复后重新运行同一命令，结果为：

```text
PASS: ptodsl-test :: sim/vmi_vmula/vmi_vmula_f32_vl1_g1_a1_n_nomask

精度校验成功:
mere=0
mare=0
```

最新成功运行的 OPPROF：

```text
build/lit-res/sim/cases/vmi_vmula/
  vmi_vmula_f32_vl1_g1_a1_n_nomask/
  OPPROF_20260902131158_RAKCGEWWPRRFVLMZ/
```

修复后的寄存器值正确：

```text
RV_VLDI -> V0 = 0x408567c4  // acc
RV_VLDI -> V1 = 0xc08842e3  // lhs
RV_VLDI -> V2 = 0xbf98a634  // rhs
```

执行：

```text
RV_VMULA Dtype:F32
Binary: 0x801047c1
PC:     0x10d0d110
```

写回结果：

```text
V0 = 0x4113f410 = 9.247086
```

与 FP32 native reference 以及 FP64 golden 在测试阈值内一致。

## 5. 最终结论

各层检查结果如下：

| 层级 | 结论 |
|---|---|
| 测试输入生成 | 正确 |
| PTODSL Python 前端 | 正确 |
| PTO MLIR `vmi.vmula` 操作数顺序 | 正确 |
| device-only LLVM intrinsic 参数顺序 | 正确 |
| Bisheng 生成的 `VLDI/VMULA/VSTI` 寄存器映射 | 正确 |
| CA model 对 `RV_VMULA` 的执行 | 正确 |
| 测试 runtime 输入 tensor 生命周期 | **错误，已修复** |

最终根因是测试框架只保留 device tensor 的裸指针，没有保留 tensor 所有者，造成 NPU 内存释放和复用。该问题使 `acc_ptr` 在运行时读到 `rhs` 数据，形成了“accumulator 被绑定为 rhs”的假象。

因此，当前复现不能证明 CANN 9.0.0 Bisheng 后端存在 `vmula` accumulator lowering 错误；修复测试框架的 device tensor 生命周期后，同一 kernel、同一 Bisheng 和同一 CA model 已通过精度验证。

**Comment by chenben09** [comment]
result = pto.vmi.vmula(acc, lhs, rhs, predicate, pmode=pmode)

浮点 dtype（f16/bf16/f32）全 VL 编译运行正常；但整型 dtype（i8/i16/i32）在全部 VL 档位
（1/8/64/128/256）、masked/unmasked 全场景编译失败（72/72 用例），且报错随 dtype 与 VL
组合分为三类：

| dtype＼VL | vl1 | vl8 | vl64 | vl128 | vl256(全部 mask 组合) |
|---|---|---|---|---|---|
| i8 | ① 粒度 | ① 粒度 | ① 粒度 | ① 粒度 | ① 粒度 |
| i16 | ① 粒度 | ① 粒度 | ① 粒度 | ① 粒度 | ① 粒度 |
| i32 | ③ 同型校验 | ④ mask layout | ② RESIDUAL | ② RESIDUAL | ② RESIDUAL |

① `'pto.vmi.vmula' op requires mask granularity to match data element width`
   （i8/i16 共 48 例，无条件触发）emission 校验要求掩码粒度与元素位宽一致；
   前端 create_mask 仅产 mask<Nxpred>(1-bit 粒度)，对 8/16-bit 元素必然违约。
   注意：unmasked 用例前端同样显式构造 all-active predicate，无法绕开。

② `VMI-RESIDUAL-OP: failed to convert all VMI ops/types to VPTO`
   （i32 vl64/vl128/vl256 共 20 例）VMI→VPTO 转换无法 legalize，报错指向 kernel 中
   首个残留 op（如 vload<vreg<128xi32>>）——i32 多寄存器场景的转换 pattern 缺失。

③ `'pto.vmi.vmula' op requires acc, lhs, rhs, and result to have identical VMI vreg types`
   （i32 vl1，2 例）前端 IR 中四个操作数均为 vreg<1xi32> 同型，emission 管线
   （布局分配/掩码改写，1 元素走 slots=8 sub-physical 布局）改写后类型分叉，
   反过来违反 vmula 自身的同型契约。

④ `'pto.vmi.vmula' op requires mask layout to match data layout`
   （i32 vl8，2 例）掩码(8xpred)经布局分配后与数据(vreg<8xi32>) layout 不一致。

### Reproduction (minimal)

carrier kernel 与 #1374 相同（load×3 → create_mask → vmula → store，完整定义见
测试仓 tools/ptodsl-test/ops/vmi_vmula.py），仅 dtype/参数不同：

```markdown
参数矩阵：dtype ∈ {i8, i16, i32}；vl ∈ {1, 8, 64, 128, 256}；
masked=true(prefix: a=0/1/VL-1/VL; group_prefix: g∈{2,4,8}, a=0/1/half) 与 false。
代表用例（测试框架可直接复现）：
  ① vmi_vmula_i8_vl1_g1_a1_n_nomask
     vmi_vmula_i16_vl128_g1_a128_n_nomask
  ② vmi_vmula_i32_vl64_g1_a64_n_nomask
     vmi_vmula_i32_vl128_g1_a128_n_nomask
     vmi_vmula_i32_vl256_g1_a256_n_nomask
  ③ vmi_vmula_i32_vl1_g1_a1_n_nomask
  ④ vmi_vmula_i32_vl8_g1_a8_n_nomask_t01

kernel.mlir 关键片段（i16 vl128，报错位置 40:12）：
%7 = pto.vmi.create_mask %c128 : index -> !pto.vmi.mask<128xpred>
%8 = pto.vmi.vmula %4, %5, %6, %7 : !pto.vmi.vreg<128xi16>, !pto.vmi.vreg<128xi16>,
     !pto.vmi.vreg<128xi16>, !pto.vmi.mask<128xpred> -> !pto.vmi.vreg<128xi16>
Expected behavior
整型 dtype 编译通过：VMI→VPTO 转换完整、掩码粒度/布局契约满足，Simulator 数值正确。
Actual behavior / error logs
# ① i8/i16 全 VL（48 例）
loc("8"("kernel.mlir":40:12)): error: 'pto.vmi.vmula' op requires mask granularity to match data element width
Error: VPTO emission pipeline failed.

# ② i32 vl64/128/256（20 例）
loc("4"("kernel.mlir":33:12)): error: VMI-RESIDUAL-OP: failed to convert all VMI ops/types to VPTO
Error: VPTO emission pipeline failed.

# ③ i32 vl1（2 例）
loc("8"("kernel.mlir":40:12)): error: 'pto.vmi.vmula' op requires acc, lhs, rhs, and result to have identical VMI vreg types

# ④ i32 vl8（2 例）
loc("8"("kernel.mlir":40:12)): error: 'pto.vmi.vmula' op requires mask layout to match data layout
失败合计 72/144 用例（i8 24 + i16 24 + i32 24）；浮点 72 例中 27 例编译运行通过，
45 例属另一运行期问题（inactive lane zeroing），另行描述。

**Comment by TelGome** [comment]
https://gitcode.com/cann/pto-as/pull/49 

剩余28 cases 模拟器报错：simulator 不支持 S8 类型的 RV_VMULA （模拟器报错，但是bisheng应该是支持的）
 

 ```
[INFO] Model Start Time: 2026-09-07 10:53:25
[DRVSTUB_LOG] driver_api.c:556 sendSwapBuf:swapbuf_base_addr:10000000
[DRVSTUB_LOG] driver_api.c:557 sendSwapBuf:sq:0 swapbuf_addr:10000000
[DRVSTUB_LOG] driver_api.c:556 sendSwapBuf:swapbuf_base_addr:10000000
[DRVSTUB_LOG] driver_api.c:557 sendSwapBuf:sq:1 swapbuf_addr:10000040
[DRVSTUB_LOG] driver_api.c:556 sendSwapBuf:swapbuf_base_addr:10000000
[DRVSTUB_LOG] driver_api.c:557 sendSwapBuf:sq:2 swapbuf_addr:10000080
[INFO] <ProfInit> Start profiling on kernel: vmi_vmula_vmi_vmula_i8_vl8_g1_a8_n_nomask_t01_mix_aiv
[info] [0000000938] [block_start]           : AIV, task_id=0, core_id=0, block_id=0
[error] ERROR MTE: dst_addr shound be 32byte aligned!
[error] ERROR MTE: dst_addr shound be 32byte aligned!
[error] ERROR MTE: dst_addr shound be 32byte aligned!
[error] [core0.veccore0] [mte_instr_addr_misalign_t0] (PC: 0x10d0d068) MTE2     : (Binary: 0x74822214) (ID: 000026) MOV_SRC_TO_DST_ALIGNv2 
[error] ERROR MTE: dst_addr shound be 32byte aligned!
[error] [core0.veccore0] [mte_instr_addr_misalign_t0] (PC: 0x10d0d070) MTE2     : (Binary: 0x74843214) (ID: 000028) MOV_SRC_TO_DST_ALIGNv2 
[ERROR] rvec_dec.cc:1047 check_addr_aligned ISA: RV_VLDI, PC: 0x10d0d108 Address 0x00000008 is not aligned to 32 bytes
Assertion failed: rvec_dec.cc:1051 check_addr_aligned
ISA: RV_VLDI, PC :0x10d0d108 Address 0x00000008 is not aligned to 32 bytes
[error] [core0.veccore0] [vec_err_instr_misalign_t0] (PC: 0x10d0d108) RVECLD   : (Binary: 0x021003d8) (ID: 000055) RV_VLDI 
[ERROR] rvec_dec.cc:1047 check_addr_aligned ISA: RV_VLDI, PC: 0x10d0d10c Address 0x00000010 is not aligned to 32 bytes
Assertion failed: rvec_dec.cc:1051 check_addr_aligned
ISA: RV_VLDI, PC :0x10d0d10c Address 0x00000010 is not aligned to 32 bytes
[error] [core0.veccore0] [vec_err_instr_misalign_t0] (PC: 0x10d0d10c) RVECLD   : (Binary: 0x041803d8) (ID: 000056) RV_VLDI 
[ERROR] pem_rvec_ooo.cc:555 RVecUop Failed to find VALU type (Unsupported Instr/Type). instr.name=RV_VMULA, type.S8, instr.id=57 instr.pc=0x10D0D110 instr.bin=0x801044C1
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
/tmp/ptodsl-test/msprof/vmi_vmula_launch.0L7tEb/run_example.sh: line 3: 188865 Aborted                 (core dumped) "${PTOAS_SIM_DSL_PYTHON_BIN}" "${PTOAS_SIM_DSL_EXAMPLE_PATH}" "$@"
2026-09-07 10:53:45 [WARN]  Child process exited with status 34304
2026-09-07 10:53:47 [WARN]  Profiling running finished. May cause generating performance files to fail.
2026-09-07 10:53:47 [WARN]  Running task failed, data parsing start
2026-09-07 10:53:47 [INFO]  Start parse dump file
2026-09-07 10:53:48 [WARN]  Failed to get any available dump file to parse, please check
2026-09-07 10:53:48 [ERROR] Parse kernel data failed, kernel name is vmi_vmula_vmi_vmula_i8_vl8_g1_a8_n_nomask_t01_mix_aiv, the order is 0
2026-09-07 10:53:48 [INFO]  Profiling results saved in /tmp/ptodsl-test/msprof/vmi_vmula_launch.0L7tEb/OPPROF_20260907105311_FIDLAELGSERDSVMD
2026-09-07 10:53:48 [WARN]  Profiling kernels result is: 0 success, 1 failed. Please check
2026-09-07 10:53:48 [ERROR] Profiling data parse failed. Please check
2026-09-07 10:53:48 [INFO]  Op profiling finish. Welcome to next use.
```
